{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 1: Review Notes/Folder\n",
    "\n",
    "Overall Goal: Examine whether genomic signatures of climate adaptation can be detected across a variety of species\n",
    "\n",
    "Method: Align sequence data from populations of Acropora digitifera, Acropora millepora, Acropora tenuis, and Acropora cervicornis to the Acropora millepora genome to identify SNPs with associations to 38 years of temperature anomaly data\n",
    "\n",
    "Bay Lab Goal: Using BAMs from the alignment, use ANGSD to Calculate per-site thetas. **What is the expectation across populations and across species?** \n",
    "\n",
    "\n",
    "### Step 2: Need to get parameters figured out for sliding window analysis\n",
    "\n",
    "##### Day 1 Goal: Run analysis with default parameters to get oriented.\n",
    "    - popgen.dk/angsd/index.php/Thetas,Tajima,Neutrality_tests\n",
    "    \n",
    "- Download NGS tools \n",
    "   ```\n",
    "   $ cd Programs\n",
    "   Programs$ interact --egress\n",
    "   Programs$ git clone https://github.com/ncbi/ngs-tools.git\n",
    "   Programs$ cd ngs-tools\n",
    "   ngs-tools$ ./configure --help  # configure in this directory.\n",
    "   ngs-tools$ ./configure --prefix=/$SCRATCH/Programs/ngs-tools\n",
    "   .\n",
    "   .\n",
    "   configure: error: required ngs-sdk package not found.\n",
    "   \n",
    "   ```\n",
    "- Based on the error, I went here (https://github.com/ncbi/sra-tools/wiki/Building-and-Installing-from-Source)and found that I need to download ncbi-vdb. \n",
    "\n",
    "    ```\n",
    "    Programs$ git clone https://github.com/ncbi/ncbi-vdb.git\n",
    "    Programs$ cd ncbi-vdb/\n",
    "    ncbi-vdb$ ./configure --prefix=/$SCRATCH/Programs/ncbi-vdb/\n",
    "    .\n",
    "    .\n",
    "    error: required ngs-sdk package not found.\n",
    "    ```\n",
    "- More error correcting. Found instructions here : https://github.com/ncbi/ngs-tools/issues/3\n",
    "    \n",
    "    ```\n",
    "    Programs$ git clone https://github.com/ncbi/ngs.git\n",
    "    Programs$ cd ngs/ngs-sdk\n",
    "    ngs-sdk]$ ./configure --prefix=/$SCRATCH/Programs/ngs/ngs-sdk && make\n",
    "    ngs-sdk]$ cd ../../ncbi-vdb\n",
    "    ncbi-vdb]$ ./configure -p=/$SCRATCH/Programs/ncbi-vdb/ && make\n",
    "    ncbi-vdb]$ cd ../ngs-tools\n",
    "    ngs-tools$ ./configure --prefix=/$SCRATCH/Programs/ngs-tools\n",
    "    ngs-tools$ exit\n",
    "    ```   \n",
    "- Download ANGSD\n",
    "\n",
    "```\n",
    "$ cd Programs\n",
    "Programs$ interact --egress\n",
    "Programs$ git clone https://github.com/samtools/htslib.git\n",
    "Programs$ git clone https://github.com/ANGSD/angsd.git \n",
    "Programs$ cd htslib;make;cd ../angsd ;make HTSSRC=../htslib\n",
    "```\n",
    "Yay! Programs have been configured and downloaded.\n",
    "Versions:\n",
    "angsd version: 0.933-106-gb0d8011 (htslib: 1.11-35-g85240ba) build(Jan 11 2021 18:17:14)\n",
    "\n",
    "\n",
    "- Set up symbolic links for all the bam files in /pylon5/ebz3a6p/rachbay/CoralCollab/mil\n",
    "\n",
    "    ```\n",
    "    Nt_div]$ mkdir millepora\n",
    "    Nt_div]$ cd millepora\n",
    "    millepora]$ ln -s /pylon5/ebz3a6p/rachbay/CoralCollab/mil/*.bam .\n",
    "    ```\n",
    "    \n",
    "- Break down Rachael's code to run my own.\n",
    "    - Note that the following code has my comments integrated using ```#[lg]```\n",
    "\n",
    "```\n",
    "#!/bin/bash \n",
    "#\n",
    "#all commands that start with SBATCH contain commands that are just used by SLURM for scheduling  \n",
    "#################\n",
    "#set a job name  \n",
    "#SBATCH --job-name=THETA\n",
    "#################  \n",
    "#a file for job output, you can check job progress\n",
    "#SBATCH --output=THETA.%j.out\n",
    "#################\n",
    "# a file for errors from the job\n",
    "#SBATCH --error=THETA.%j.err\n",
    "#################\n",
    "#time you think you need; default is one hour\n",
    "#in minutes in this case\n",
    "#SBATCH -t 48:00:00\n",
    "#################\n",
    "#quality of service; think of it as job priority\n",
    "#SBATCH -p RM\n",
    "#################\n",
    "#number of nodes\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=28\n",
    "#################\n",
    "#SBATCH --mem=120G\n",
    "#################\n",
    "#################\n",
    "#get emailed about job BEGIN, END, and FAIL\n",
    "#SBATCH --mail-type=ALL\n",
    "#################\n",
    "#who to send email to; please change to your email\n",
    "#SBATCH  --mail-user=rachaelbay@gmail.com\n",
    "#################\n",
    "#now run normal batch commands\n",
    "##################\n",
    "#echo commands to stdout\n",
    "\n",
    "set -x\n",
    "\n",
    "#[lg] variables for ref genome and where dependencies(?) are\n",
    "REFERENCE=$SCRATCH/CoralMeta/Amil.v2.01.chrs.fasta   \n",
    "NGSTOOLS=/home/rachbay/programs/ngsTools\n",
    "\n",
    "#[lg] variable for the species being worked on\n",
    "pref='mil'\n",
    "\n",
    "mkdir thetas/$pref\n",
    "\n",
    "\n",
    "\n",
    "    #~/programs/angsd/angsd -bam milbam.txt -P 2 -ref $REFERENCE -anc $REFERENCE \\\n",
    "    #       -remove_bads 1 -uniqueOnly 1 -only_proper_pairs -minInd 20 -doMaf 1 -doMajorMinor 4 \\\n",
    "    #       -doCounts 1 -GL 1 -doSaf 1 -out thetas/$pref/$pref\n",
    "\n",
    "#[lg] This code is for allele frequency estimation from genotype likelihoods with bam files \n",
    "#[lg] The the site frequency spectrum is also estimated here. \n",
    "#[lg] milbam.txt is a list of the bam files \n",
    "#[lg] -P............................number of threads allocated to the program\n",
    "#[lg] -ref..........................path to reference genome\n",
    "#[lg] -anc..........................path to fasta file with the ancestral alleles, can be null\n",
    "#[lg] -remove_bads..................Discard 'bad' reads, (flag >=256) \n",
    "#[lg] -uniqueOnly...................Discards reads that don't map uniquely\n",
    "#[lg] -only_proper_pairs............Only use reads where the mate could be mapped\n",
    "#[lg] -minInd.......................Discard site if effective sample size below value. (based on read depth)\n",
    "#[lg] -doMaf........................Estimate allele frequencies, 1: Frequency (fixed major and minor)\n",
    "#[lg] -doMajorMinor.................Infer the major/minor using different approaches, 4:Use reference allele as major (requires -ref) \n",
    "#[lg] -doCounts.....................Count the number A,C,G,T. All sites, All samples\n",
    "#[lg] -GL...........................Estimate genotype likelihoods, 1:SAMtools\n",
    "#[lg] -doSaf........................Estimate the SFS and/or neutrality tests genotype calling, 1: perform multisample GL estimation\n",
    "#[lg] -out..........................Out Dir\n",
    "#[lg] which option requires the -anc\n",
    "\n",
    "\n",
    "\n",
    "~/programs/angsd/misc/realSFS thetas/$pref/$pref.saf.idx -P 1 -fold 1 -nSites 50000000 > thetas/$pref/$pref.sfs\n",
    "~/programs/angsd/misc/realSFS saf2theta thetas/$pref/$pref.saf.idx -outname thetas/$pref/$pref -sfs thetas/$pref/$pref.sfs -fold 1\n",
    "\n",
    "~/programs/angsd/misc/thetaStat do_stat thetas/$pref/$pref.thetas.idx\n",
    "~/programs/angsd/misc/thetaStat print thetas/$pref/$pref.thetas.idx > thetas/$pref/$pref.thetas\n",
    "~/programs/angsd/misc/thetaStat do_stat thetas/$pref/$pref.thetas.idx -win 1000 -step 500 -outnames thetas/$pref/$pref.thetasWindow.gz\n",
    "\n",
    "\n",
    "    \n",
    "    ```\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
